src/NetTopologySuite/Geometries/Geometry.cs

```diff
@@ -1515,24 +1517,7 @@ public virtual Geometry Reverse()
         /// <exception cref="ArgumentException">if the argument is a non-empty heterogeneous <c>GeometryCollection</c></exception>
         public Geometry Intersection(Geometry other)
         {
-            // Special case: if one input is empty ==> empty
-            if (IsEmpty || other.IsEmpty)
-                return OverlayOp.CreateEmptyResult(SpatialFunction.Intersection, this, other, _factory);
-
-            // compute for GCs
-            // (An inefficient algorithm, but will work)
-            // TODO: improve efficiency of computation for GCs
-            if (IsGeometryCollection)
-            {
-                var g2 = other;
-                return GeometryCollectionMapper.Map(
-                    (GeometryCollection)this, g => g.Intersection(g2));
-            }
-
-            // No longer needed since GCs are handled by previous code
-            //CheckNotGeometryCollection(this);
-            //CheckNotGeometryCollection(other);
-            return SnapIfNeededOverlayOp.Overlay(this, other, SpatialFunction.Intersection);
+            return Factory.GeometryOverlay.Intersection(this, other);
```

Comment:
I'd like it if this (and the other similar instance methods) were to check and throw an exception if `!Equals(Factory.GeometryOverlay, other.Factory.GeometryOverlay)`.

This isn't a problem in JTS, since there's only one overlay implementation that's allowed to exist in the entire package at any given time, but for us, we **do** allow there to be more than one, so it could happen.

It's my opinion that `a.Intersection(b)`, `a.Union(b)`, and `a.SymmetricDifference(b)` should always behave practically identically to `b.Intersection(a)`, `b.Union(a)`, and `b.SymmetricDifference(a)` (plus-or-minus minor differences related to numerical robustness).  As-written, there could be substantial differences between the two, because they could use significantly different algorithms, which could perhaps surface as a "gotcha" in some use cases that we don't know about at the moment, which could add another layer of difficulty when debugging.

As for `Difference`, I don't have a specific reason to want it to throw an exception if the `GeometryOverlay` implementations are different, other than consistency with the other bivariate overlay operations.

Reply:
Equality check in `GeometryOverlay`